plugins/signals-backend/src/service/router.ts

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2023 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { errorHandler } from '@backstage/backend-common';
+import express, { NextFunction, Request, Response } from 'express';
+import Router from 'express-promise-router';
+import { Logger } from 'winston';
+import { SignalService } from '@backstage/plugin-signals-node';
+import * as https from 'https';
+import http from 'http';
+
+/** @public */
+export interface RouterOptions {
+  logger: Logger;
+  service: SignalService;
+}
+
+/** @public */
+export async function createRouter(
+  options: RouterOptions,
+): Promise<express.Router> {
+  const { logger, service } = options;
+  let subscribed = false;
+  const upgradeMiddleware = (req: Request, _: Response, next: NextFunction) => {
+    const server: https.Server | http.Server = (req.socket as any)?.server;
+    if (
+      !server ||
+      !req.headers ||
+      req.headers.upgrade === undefined ||
+      req.headers.upgrade.toLowerCase() !== 'websocket'
+    ) {
+      next();
+      return;
+    }
+
+    if (!subscribed) {
+      subscribed = true;
+      server.on('upgrade', async (request, socket, head) => {
+        // Only upgrade if request to root of the signals plugin
+        if (request.url === '/api/signals') {
```

Comment:
Could you add a big TODO here to figure out how to make this more generic? We can't assume that this is the path that the plugin is mounted at. It might actually be that the kubernetes backend already has a solution to this as well.

Reply:
Yeah absolutely. I couldn't figure out a way to see the mounted route so had to hardcode it for now. Maybe something for the future.